File: src/main/java/it/unibo/crabinv/model/save/PlayerBaseStats.java
package it.unibo.crabinv.model.save;

import it.unibo.crabinv.model.powerups.PowerUpType;

public enum PlayerBaseStats {
    SPEED(0.01, PowerUpType.SPEED_UP),
    FIRE_RATE(8, PowerUpType.FIRERATE_UP),
    PLAYER_HEALTH(2, PowerUpType.HEALTH_UP);

    private final double value;
    private final PowerUpType powerUpType;

    PlayerBaseStats(double value, PowerUpType powerUpType) {
        this.value = value;
        this.powerUpType = powerUpType;
    }

    public int getIntValue() {
        return (int) this.value;
    }

    public double getDoubleValue() {
        return this.value;
    }

    public PowerUpType getPowerUpType(){
        return this.powerUpType;
    }

    public static double getDoubleValueOf(PowerUpType powerUpType){
        for (PlayerBaseStats stat : values()){
            if (stat.powerUpType == powerUpType){
                return stat.getDoubleValue();
            }
        }
        throw new IllegalArgumentException("PowerUpType " + powerUpType + " is not associated to any PlayerBaseStat");
    }

    public static int getIntValueOf(PowerUpType powerUpType){
        return (int) getDoubleValueOf(powerUpType);
    }

}

File: src/main/java/it/unibo/crabinv/model/core/save/PlayerBaseStats.kt
package it.unibo.crabinv.model.core.save

import it.unibo.crabinv.model.powerups.PowerUpType

/**
 * `enum` of all the admissible player stats.
 */
enum class PlayerBaseStats(
    /**
     * @return the `double` value selected
     */
    val doubleValue: Double,
    private val powerUpType: PowerUpType,
) {
    SPEED(0.01, PowerUpType.SPEED_UP),
    FIRE_RATE(8.0, PowerUpType.FIRERATE_UP),
    PLAYER_HEALTH(2.0, PowerUpType.HEALTH_UP), ;

    companion object {
        /**
         * Returns the `double` value to [Int] of the value associated with the `PowerUpType`.
         * @param powerUpType the `PowerUpType` to look up
         * @return the associated `double` value
         */
        @JvmStatic
        fun getDoubleValueOf(powerUpType: PowerUpType): Double {
            for (stat in entries) {
                if (stat.powerUpType == powerUpType) {
                    return stat.doubleValue
                }
            }
            throw IllegalArgumentException("PowerUpType $powerUpType is not associated to any PlayerBaseStat")
        }

        /**
         * Returns the value cast to [Int] of the value associated with the `PowerUpType`.
         * @param powerUpType the `PowerUpType` to look up
         * @return the associated value cast to [Int]
         */
        fun getIntValueOf(powerUpType: PowerUpType): Int = getDoubleValueOf(powerUpType).toInt()
    }
}
